### PR TITLE
Added delegate functionality for ReadToken() into JwtSecurityTokenHan…

### LINKED ISSUE #798: Add delegate for ReadToken similar to SignatureValidator
Customers that have legacy code or who want to apply different configuration for tokens could use this delegate to read the token once, obtain the correct policy and proceed with validation.

### PR DESCRIPTION
**Pull request by mafurman**
Fixes #798.

### CODE REVIEW

File: src/Microsoft.IdentityModel.Tokens/TokenValidationParameters.cs [hunk 1]
```diff
@@ -96,6 +96,12 @@ namespace Microsoft.IdentityModel.Tokens
     public delegate SecurityToken SignatureValidator(string token, TokenValidationParameters validationParameters);
 
     /// <summary>
+    /// Definition for TokenReader.
+    /// </summary>
+    /// <param name="token">A securityToken with a signature.</param>
+    public delegate SecurityToken TokenReader(string token);
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/TokenValidationParameters.cs]: Even though it is not needed, we always pass TokenValidationParameters to facilitate extensibility.

File: src/Microsoft.IdentityModel.Tokens/TokenValidationParameters.cs [hunk 2]
```diff
@@ -96,6 +96,12 @@ namespace Microsoft.IdentityModel.Tokens
...
+
+    /// <summary>
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/TokenValidationParameters.cs]: In the ctor: TokenValidationParameters(TokenValidationParameters other), we need to set TokenReader to other.TokenReader

File: src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs [hunk 3]
```diff
@@ -931,7 +931,20 @@ protected virtual JwtSecurityToken ValidateSignature(string token, TokenValidati
...
+            if (validationParameters.TokenReader != null)
+            {
+                var securityToken = validationParameters.TokenReader(token);
+                if (securityToken == null)
+                    throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(LogHelper.FormatInvariant(TokenLogMessages.IDX10507, token)));
+
+                jwtToken = securityToken as JwtSecurityToken;
+                if (jwtToken == null)
+                    throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(LogHelper.FormatInvariant(TokenLogMessages.IDX10508, typeof(JwtSecurityToken), securityToken.GetType(), token)));
+            } else
```
Comment (brentschmaltz) [inline src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs]: In this case we should put the else on a separate line and then use { }. Even though the else has only one clause.

File: test/Microsoft.IdentityModel.Tests/ValidationDelegates.cs [hunk 4]
```diff
@@ -29,6 +29,8 @@
 using System.Collections.Generic;
 using System.IdentityModel.Tokens.Jwt;
 using Microsoft.IdentityModel.Tokens;
+using Microsoft.IdentityModel.Tokens.Saml;
+using Microsoft.IdentityModel.Tokens.Saml2;
```
Comment (brentschmaltz) [inline test/Microsoft.IdentityModel.Tests/ValidationDelegates.cs]: we don't need M.IM.Tokens or M.IM.Tokens.Saml usings.
Another way to return a different token, is to define a CustomSecurityToken like in: CustomTypes.cs. If we use CustomSecurityToken, then we can remove the M.IM.Tokens.Saml2 using.

In fact, this CustomSecurityToken should be moved to DerivedTypes.cs.

File: src/Microsoft.IdentityModel.Tokens/TokenValidationParameters.cs (hunk 1 above)
Comment (mafurman) [reply src/Microsoft.IdentityModel.Tokens/TokenValidationParameters.cs]: Got it!

File: src/Microsoft.IdentityModel.Tokens/TokenValidationParameters.cs (hunk 2 above)
Comment (mafurman) [reply src/Microsoft.IdentityModel.Tokens/TokenValidationParameters.cs]: Sorry for missing this. Just fixed it.

File: src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs (hunk 3 above)
Comment (mafurman) [reply src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs]: Fixed it! I'll try to be more careful next time!

File: test/Microsoft.IdentityModel.Tests/ValidationDelegates.cs (hunk 4 above)
Comment (mafurman) [reply test/Microsoft.IdentityModel.Tests/ValidationDelegates.cs]: Moved the CustomSecurityToken into DerivedTypes.cs and removed the unnecessary using statements.

File: src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs [hunk 5]
```diff
@@ -948,7 +948,22 @@ protected virtual JwtSecurityToken ValidateSignature(string token, TokenValidati
...
+            if (validationParameters.TokenReader != null)
+            {
+                var securityToken = validationParameters.TokenReader(token, validationParameters);
+                if (securityToken == null)
+                    throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(LogHelper.FormatInvariant(TokenLogMessages.IDX10510, token)));
+
+                jwtToken = securityToken as JwtSecurityToken;
+                if (jwtToken == null)
+                    throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(LogHelper.FormatInvariant(TokenLogMessages.IDX10509, typeof(JwtSecurityToken), securityToken.GetType(), token)));
+            } else
```
Comment (brentschmaltz) [inline src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs]: please put else on separate line.

**Comment by brentschmaltz** [comment]
one small comment then :shipit:

**Pull request merged**